File: trustbit_mandi/trustbit_mandi/report/soda_price_list_ledger/soda_price_list_ledger.py
# Copyright (c) 2026, Trustbit Software and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	return columns, data


def get_columns():
	return [
		{"fieldname": "sno", "label": _("S.No."), "fieldtype": "Int", "width": 50},
		{"fieldname": "effective_datetime", "label": _("Effective Date Time"), "fieldtype": "Datetime", "width": 160},
		{"fieldname": "price_list_area", "label": _("Area"), "fieldtype": "Link", "options": "Soda Price List Area", "width": 120},
		{"fieldname": "item", "label": _("Item"), "fieldtype": "Link", "options": "Item", "width": 150},
		{"fieldname": "item_group", "label": _("Item Group"), "fieldtype": "Data", "width": 120},
		{"fieldname": "base_price_50kg", "label": _("Base Price (50KG)"), "fieldtype": "Currency", "width": 130},
		{"fieldname": "price_per_kg", "label": _("Price/KG"), "fieldtype": "Currency", "width": 100},
		{"fieldname": "is_active", "label": _("Active"), "fieldtype": "Check", "width": 60},
		{"fieldname": "remarks", "label": _("Remarks"), "fieldtype": "Data", "width": 150},
		{"fieldname": "name", "label": _("ID"), "fieldtype": "Link", "options": "Soda Price List", "width": 140},
	]


def get_data(filters):
	conditions = ["1=1"]
	values = []

	if filters.get("price_list_area"):
		conditions.append("price_list_area = %s")
		values.append(filters["price_list_area"])

	if filters.get("item"):
		conditions.append("item = %s")
		values.append(filters["item"])

	if filters.get("item_group"):
		conditions.append("item_group = %s")
		values.append(filters["item_group"])

	if filters.get("from_date"):
		conditions.append("effective_datetime >= %s")
		values.append(filters["from_date"])

	if filters.get("to_date"):
		conditions.append("effective_datetime <= %s")
		values.append(filters["to_date"] + " 23:59:59")

	sql = """
		SELECT name, effective_datetime, price_list_area, item, item_group,
		       base_price_50kg, price_per_kg, is_active, remarks
		FROM `tabSoda Price List`
		WHERE {conditions}
		ORDER BY price_list_area ASC, item ASC, effective_datetime DESC
	""".format(conditions=" AND ".join(conditions))

	data = frappe.db.sql(sql, values, as_dict=True)

	sno = 0
	for row in data:
		sno += 1
		row["sno"] = sno

	return data

File: trustbit_mandi/trustbit_mandi/report/soda_price_list_ledger/soda_price_list_ledger.js
// Copyright (c) 2026, Trustbit Software and contributors
// For license information, please see license.txt

frappe.query_reports["Soda Price List Ledger"] = {
	onload: function(report) {
		report.page.add_inner_button(__("Print PDF"), function() {
			print_report_pdf(report, "Soda Price List Ledger");
		});
	},
	"filters": [
		{
			"fieldname": "price_list_area",
			"label": __("Price List Area"),
			"fieldtype": "Link",
			"options": "Soda Price List Area"
		},
		{
			"fieldname": "item",
			"label": __("Item"),
			"fieldtype": "Link",
			"options": "Item"
		},
		{
			"fieldname": "item_group",
			"label": __("Item Group"),
			"fieldtype": "Link",
			"options": "Item Group"
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1)
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today()
		}
	]
};

function print_report_pdf(report, title) {
	let filters = report.get_values();
	let filter_text = [];
	if (filters.from_date) filter_text.push("From: " + filters.from_date);
	if (filters.to_date) filter_text.push("To: " + filters.to_date);
	Object.keys(filters).forEach(function(k) {
		if (k !== "from_date" && k !== "to_date" && filters[k]) {
			filter_text.push(k.replace(/_/g, " ").replace(/\b\w/g, c => c.toUpperCase()) + ": " + filters[k]);
		}
	});

	let data_table = document.querySelector(".dt-scrollable") || document.querySelector(".datatable");
	if (!data_table) {
		frappe.msgprint(__("No report data to print. Please run the report first."));
		return;
	}
	let table_html = data_table.querySelector("table")
		? data_table.querySelector("table").outerHTML
		: data_table.innerHTML;

	let w = window.open();
	w.document.write('<html><head><title>' + title + '</title>');
	w.document.write('<style>');
	w.document.write('@page { size: A4 landscape; margin: 8mm; }');
	w.document.write('body { font-family: Arial, sans-serif; font-size: 11px; }');
	w.document.write('.header { text-align: center; margin-bottom: 10px; }');
	w.document.write('.header h2 { margin: 0; font-size: 16px; }');
	w.document.write('.header p { margin: 3px 0; font-size: 11px; color: #555; }');
	w.document.write('table { width: 100%; border-collapse: collapse; }');
	w.document.write('th, td { border: 1px solid #333; padding: 4px 6px; font-size: 10px; }');
	w.document.write('th { background: #e5e5e5; font-weight: bold; }');
	w.document.write('td { text-align: right; }');
	w.document.write('td:first-child, td:nth-child(2), td:nth-child(3) { text-align: left; }');
	w.document.write('.footer { margin-top: 10px; font-size: 9px; color: #888; text-align: right; }');
	w.document.write('</style></head><body>');
	w.document.write('<div class="header">');
	w.document.write('<h2>' + title + '</h2>');
	if (filter_text.length) w.document.write('<p>' + filter_text.join(" | ") + '</p>');
	w.document.write('</div>');
	w.document.write(table_html);
	w.document.write('<div class="footer">Printed on: ' + frappe.datetime.now_datetime() + '</div>');
	w.document.write('</body></html>');
	w.document.close();
	setTimeout(function() { w.print(); }, 500);
}
